### PR TITLE
Add Mixture Of Agents paper implementation

### PR DESCRIPTION
**Pull request by ravi03071991**
# Description

PR to add [Mixture-Of-Agents](https://arxiv.org/abs/2406.04692) paper as LlamaPack.

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        self,
+        llm: LLM,
+        reference_llms: List[LLM],
+        num_layers: int = 3,
+    ) -> None:
+        self.llm = llm
+        self.reference_llms = reference_llms
+        self.num_layer = num_layers
+
+    def inject_references_to_messages(
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py:25]: missing a few types and return types on a few function definitions like this one

File: llama-index-packs/llama-index-packs-mixture-of-agents/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Mixture-Of-Agents Pack
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-mixture-of-agents/README.md:1]: I wonder if we should write a disclaimer on the name lol -- people might be confused

File: llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    ):
+        if len(references) > 0:
+            messages = self.inject_references_to_messages(messages, references)
+
+        return str(llm.chat(messages, max_tokens=max_tokens, temperature=temperature))
+
+    def get_answer(self, query_str: str):
+        messages = []
+
+        messages.append(ChatMessage(role="system", content=query_str))
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py:63]: should this be user role?

File: llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+            for layer in range(self.num_layers):
+                logger.info(
+                    f"Round {layer+1}/{self.num_layers} to collecting reference responses."
+                )
+
+                references = []
+
+                for reference_llm in self.reference_llms:
+                    reference = self.generate_with_references(
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py:78]: probably much faster to use async with this. Basically, make `generate_with_references()` async and have it use `await llm.achat()`

And then

```python
jobs = [self.agenerate_with_references(...)]
references = await asyncio.gather(*jobs)
```

Comment (logan-markewich) [reply llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py:78]: Purely optional, but would make it a lot more efficient

File: llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py (hunk 3 above)
Comment (ravi03071991) [reply llama-index-packs/llama-index-packs-mixture-of-agents/llama_index/packs/mixture_of_agents/base.py:63]: Yeah, should be user. Got confused.

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**